tiledb/sm/filesystem/vfs.cc

```diff
@@ -811,66 +812,106 @@ Status VFS::terminate() {
   }
 
-  std::vector<std::string> paths;
+  optional<std::vector<FileStat>> entries;
   if (parent.is_file()) {
 #ifdef _WIN32
-    RETURN_NOT_OK(win_.ls(parent.to_path(), &paths));
-#else
-    RETURN_NOT_OK(posix_.ls(parent.to_path(), &paths));
-#endif
-  } else if (parent.is_hdfs()) {
-#ifdef HAVE_HDFS
-    RETURN_NOT_OK(hdfs_->ls(parent, &paths));
+    Status st;
+    std::tie(st, entries) = win_.ls_with_sizes(parent);
 #else
-    return LOG_STATUS(Status_VFSError("TileDB was built without HDFS support"));
+    Status st;
+    std::tie(st, entries) = posix_.ls_with_sizes(parent);
 #endif
+    if (!st.ok()) {
+      return {st, nullopt};
+    }
   } else if (parent.is_s3()) {
 #ifdef HAVE_S3
-    RETURN_NOT_OK(s3_.ls(parent, &paths));
+    Status st;
+    std::tie(st, entries) = s3_.ls_with_sizes(parent);
 #else
-    return LOG_STATUS(Status_VFSError("TileDB was built without S3 support"));
+    auto st =
+        LOG_STATUS(Status_VFSError("TileDB was built without S3 support"));
 #endif
+    if (!st.ok()) {
+      return {st, std::nullopt};
+    }
   } else if (parent.is_azure()) {
 #ifdef HAVE_AZURE
-    RETURN_NOT_OK(azure_.ls(parent, &paths));
+    Status st;
+    std::tie(st, entries) = azure_.ls_with_sizes(parent);
 #else
-    return LOG_STATUS(
-        Status_VFSError("TileDB was built without Azure support"));
+    auto st =
+        LOG_STATUS(Status_VFSError("TileDB was built without Azure support"));
 #endif
+    if (!st.ok()) {
+      return {st, std::nullopt};
+    }
   } else if (parent.is_gcs()) {
 #ifdef HAVE_GCS
-    RETURN_NOT_OK(gcs_.ls(parent, &paths));
+    Status st;
+    std::tie(st, entries) = gcs_.ls_with_sizes(parent);
 #else
-    return LOG_STATUS(Status_VFSError("TileDB was built without GCS support"));
+    auto st =
+        LOG_STATUS(Status_VFSError("TileDB was built without GCS support"));
 #endif
+    if (!st.ok()) {
+      return {st, std::nullopt};
+    }
+  } else if (parent.is_hdfs()) {
+#ifdef HAVE_HDFS
+    Status st;
+    std::tie(st, entries) = hdfs_->ls_with_sizes(parent);
+#else
+    auto st =
+        LOG_STATUS(Status_VFSError("TileDB was built without HDFS support"));
+#endif
+    if (!st.ok()) {
+      return {st, std::nullopt};
+    }
   } else if (parent.is_memfs()) {
-    RETURN_NOT_OK(memfs_.ls(parent.to_path(), &paths));
-    // URI class expects paths to be prepended
-    for (std::string& path : paths) {
-      path.insert(0, "mem://");
+    Status st;
+    std::tie(st, entries) =
+        memfs_.ls_with_sizes(URI("mem://" + parent.to_path()));
+    if (!st.ok()) {
+      return {st, std::nullopt};
     }
   } else {
-    return LOG_STATUS(
+    auto st = LOG_STATUS(
         Status_VFSError("Unsupported URI scheme: " + parent.to_string()));
+    return {st, std::nullopt};
   }
-  parallel_sort(compute_tp_, paths.begin(), paths.end());
-  for (auto& path : paths) {
-    uris->emplace_back(path);
-  }
-  return Status::Ok();
+  parallel_sort(compute_tp_, entries->begin(), entries->end());
```

Comment:
Do we need this? Isn't it guaranteed that the entries will be sorted lexicographically? I know this is old code, but just wanted to make sure it's necessary.

Reply:
Thanks for the review @stavrospapadopoulos!
I don't think we have a guarantee that this happens. I'll give you an example, but let me repeat myself, I'm not sure this is 100% correct, it's based on my understanding of the code I read. When you list the objects in S3, the request outcome gives you a list of the objects, and a list of what S3 API calls CommonPrefixes (which I guess can be conceptually called directories). These are separate lists and currently `s3.cc` just appends the "directories" at the end of the objects list, so in this example at least, we don't return sorted entries, thus the final sort is necessary.

Reply:
Thanks, let's keep it.